### ISSUE TITLE
Loading data from ManifestArrays without saving references to disk first

### ISSUE THREAD

**Issue by ayushnag** [zarr-developers/VirtualiZarr#124]
I am working on a feature in `virtualizarr` [to read dmrpp metadata files and create a virtual `xr.Dataset`](https://github.com/TomNicholas/VirtualiZarr/issues/85) containing manifest array's that can then be virtualized. This is the current workflow: 
```python
vdatasets = parser.parse(dmrs)
# vdatasets are xr.Datasets containing ManifestArray's
mds = xr.combine_nested(list(vdatasets), **xr_combine_kwargs)
mds.virtualize.to_kerchunk(filepath=outfile, format=outformat)
ds = xr.open_dataset(outfile, engine="virtualizarr", ...)
ds.time.values
```
However the chunk manifest, encoding, attrs, etc. is already in `mds` so is it possible to read data directly from this dataset? My understanding is that once the "chunk manifest" ZEP is approved and the `zarr-python` reader in `xarray` is updated this should be possible. The `xarray` reader for `kerchunk` can accept a file or the reference json object directly from `kerchunk` `SingleHdf5ToZarr ` and `MultiZarrToZarr`. So similarly can we extract the refs from `mds` and pass it to `xr.open_dataset()` directly?

There probably still needs to be a function that extracts the refs so that xarray can make a new `Dataset` object with all the indexes, cf_time handling, and `open_dataset` checks.

```python
mds = xr.combine_nested(list(vdatasets), **xr_combine_kwargs)
refs = mds.virtualize()
ds = xr.open_dataset(refs, engine="virtualizarr", ...)
```

Even reading directly from the ManifestArray dataset is possible but not sure how the new dataset object with numpy arrays and indexes would be separate from the original dataset
```python
mds = xr.combine_nested(list(vdatasets), **xr_combine_kwargs)
mds.time.values
```

**Comment by TomNicholas** [comment]
(One subtlety I'm not sure about here would be around indexes. I think you would probably want to have a solution for loading indexes as laid out in https://github.com/zarr-developers/VirtualiZarr/issues/18, and then have the indexes understand how they can be loaded.)

**Comment by TomNicholas** [comment]
Another subtlety to consider is when should the CF decoding happen? You would then have effectively done `open_dataset` in a very roundabout way, and we need to make sure not to forget the CF decoding step in there somewhere.

**Comment by TomNicholas** [comment]
Whilst I've gone off the idea of an xarray backend that loads data by default (https://github.com/zarr-developers/VirtualiZarr/issues/221), there's another potential use case for this idea: `loadable_variables`.

Currently there are xarray backends and virtualizarr virtual backends. But the latter has to call the former if it wants to support the `loadable_variables` kwarg. For example, @sharkinsspatial 's new HDF virtual reader internally calls https://github.com/zarr-developers/VirtualiZarr/blob/1dbd119993a2f900142ee8af0ab34057f28c3831/virtualizarr/readers/hdf/hdf.py#L71

which calls `xr.open_dataset`

https://github.com/zarr-developers/VirtualiZarr/blob/1dbd119993a2f900142ee8af0ab34057f28c3831/virtualizarr/readers/common.py#L50

which for a netCDF4 file would call [xarray's `netCDF4` backend](https://github.com/pydata/xarray/blob/main/xarray/backends/netCDF4_.py) to load the `loadable_variables`. 

So to support virtualizing one filetype, we need both a virtualizarr backend and a corresponding xarray backend. For filetypes that already have xarray backends this isn't a problem, but if we want 3rd parties to add their own virtual backends for custom filetypes, it's doubling the work they have to do.

**Comment by TomNicholas** [comment]
@ayushnag I feel like there might be some sneaky trick we could do to load the `ManifestArrays` into memory using `xr.open_zarr` using either [xpublish](https://github.com/xpublish-community/xpublish) or some kind of special python zarr store implementation... 

Like what if we create some kind of special `ManifestStore` class along the lines of all the zarr stores [listed here](https://zarr.readthedocs.io/en/stable/api/storage.html), then use the zarr-python library to read from that? Is that possible?

https://zarr.readthedocs.io/en/stable/api/storage.html

**Comment by ayushnag** [comment]
Perhaps a `Store` very similar to an `IcechunkStore` except with a subset of the capabilities and requirements? It can have just the portion that provides the abstraction to read from a remote store given only the manifest

**Comment by TomNicholas** [comment]
Yes exactly! And instead of finding the byte range info from files on disk it just gets them from the ManifestArray that's in memory.

The advantage of this approach is that we should be able to then just call `open_zarr` internally to get at the loadable variables, guaranteeing that Xarray will handle CF decoding of those variables in the way it currently does when we use the netCDF backend.

**Comment by TomNicholas** [comment]
@ayushnag following your suggestion on slack, maybe what we need is to temporarily create an  icechunkstore then read from it? Though that would introduce a dependency on icechunk outside of the actual `.to_icechunk` method.

Alternatively maybe creating an `FsspecStore` (or whatever it's called exactly) that can read bytes from S3 ends up being functionally equivalent to your suggestion to use kerchunk above.

It would be cool to find a way to do this using just zarr-python.

cc @mpiannucci and @jhamman for feedback on this idea
